### PR TITLE
Fix/backend/friend

### PR DESCRIPTION
**Pull request by usatie**
- BlockされてたらFriend Request送れない
- すでにFriendになってたらFriend Request送れない
- 自分自身にはFriend Request送れない
- Friend Requestが来ていないのにAccept/Rejectできない
- Friend Requestが送ってないのにCancelできない
などを実装しました。

また、上記の機能を並行処理でもレースコンディションにならないようにロックを取ったりするためにSQLのTransactionという仕組みを使っていますので、もし興味があれば調べてみてください。

## トランザクションの簡単な説明
```
1. Confirm B is not blocked by A
2. B send friend request to A
```
というのを普通に書くと、1と2の間にblockが行われた場合におかしなことになります。なので、1と2の間にデータベースに変更がないことを保証してくれる機能がtransactionです。

出来上がりとしては、こんな感じのイメージになります。
```
1. begin transaction
2. Confirm B is not blocked by A
3. B send friend request to A
4. commit transaction
```

（Prismaがtransactionを抽象化してくれてるので、使うときはこういう見た目とは少し違いますが。）

## Summary by CodeRabbit

- **New Features**
  - Enhanced friend request functionality with improved validation and error handling.
  - Implemented transactional operations for friend request actions to ensure data integrity.

- **Bug Fixes**
  - Fixed issues where users could send friend requests to themselves or to users they are already friends with or have blocked.

- **Refactor**
  - Centralized authorization checks using `UserGuard` to streamline friend request operations.
  - Refactored friend request methods for better readability and maintainability.

- **Tests**
  - Added comprehensive end-to-end test cases for friend request edge cases.

- **Documentation**
  - Updated function parameters for clarity in test specifications.

### CODE REVIEW

File: backend/src/friend-request/friend-request.controller.ts [hunk 1]
```diff
@@ -29,78 +29,64 @@ export class FriendRequestController {
 
   // Send a friend request
   @Post()
+  @UseGuards(UserGuard)
   @UseGuards(JwtAuthGuard)
```
[the remark this answers is not in the document]
Comment (usatie) [reply backend/src/friend-request/friend-request.controller.ts:33]: こんな書き方できるのか〜〜〜！

File: backend/src/user/user.guard.ts [hunk 2]
```diff
@@ -11,6 +11,9 @@ export class UserGuard implements CanActivate {
     if (!params?.userId) {
       throw new Error('UserGuard should only be used on routes with a userId');
     }
-    return user?.id === Number(params.userId);
+    if (typeof params.userId !== 'string' || !/^\d+$/.test(params.userId)) {
+      throw new Error('userId parameter must be a valid integer');
```
[the remark this answers is not in the document]
Comment (usatie) [reply backend/src/user/user.guard.ts:15]: ParseIntPipeをGuardの中で使える？

[the remark this answers is not in the document]
Comment (usatie) [reply backend/src/user/user.guard.ts]: ok

File: backend/src/friend-request/friend-request.service.ts [hunk 3]
```diff
@@ -1,22 +1,37 @@
...
-          requesting: {
-            connect: { id: createFriendRequestDto.recipientId },
+    const { recipientId } = createFriendRequestDto;
+    if (recipientId === user.id) {
+      throw new BadRequestException('Cannot send friend request to self');
+    }
+    return this.prisma.$transaction(async (tx) => {
+      await this.expectNotRequesting(recipientId, user, tx);
+      await this.expectNotFriends(recipientId, user, tx);
+      await this.expectNotBlockedBy(recipientId, user, tx);
```
[the remark this answers is not in the document]
Comment (takumihara) [reply backend/src/friend-request/friend-request.service.ts:23]: Screenshot 2023-12-12 at 18 15 56
https://zenn.dev/tockn/articles/4268398c8ec9a9#%E3%83%88%E3%83%A9%E3%83%B3%E3%82%B6%E3%82%AF%E3%82%B7%E3%83%A7%E3%83%B3%E3%81%AE%E5%88%86%E9%9B%A2%E6%80%A7%E3%81%A8%E3%81%9D%E3%81%AE%E9%99%90%E7%95%8C
Transactionの理解がまだ甘いのですが、こういった事象って起きないのですか？

[the remark this answers is not in the document]
Comment (takumihara) [reply backend/src/friend-request/friend-request.service.ts:23]: シリアライザブル分離レベルにしないとこれ起こる気がしますね。それか、排他ロックを取るか。

https://zenn.dev/bellwood4486/articles/postgresql-trans-iso-level#%E7%9B%B4%E5%88%97%E5%8C%96%E7%95%B0%E5%B8%B8
